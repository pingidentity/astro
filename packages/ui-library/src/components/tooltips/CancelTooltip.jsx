"use strict";
var React = require("react"),
    DetailsTooltip = require("./DetailsTooltip.jsx");

/**
* @callback CancelTooltip~onCancel
*/

/**
* @callback CancelTooltip~onConfirm
*/

/**
 * @class CancelTooltip
 * @desc CancelTooltip implements tooltip used to confirm or deny the canceling of an action.
 *
 * @param {string} [data-id]
 *     The data-id of the component.
 * @param {string} cancelButtonText
 *     The text of the cancel link.
 * @param {string} confirmButtonText
 *     The text of the confirm button.
 * @param {string} messageText
 *     The text to display in the body of the tooltip.
 * @param {string} [positionClassName]
 *     CSS classes to set on the top-level HTML container. Used to manage tooltip callout positioning with the
 *     CancelTooltip.positionStyles enum and/or any extra css styling if needed.
 * @param {string} title
 *     The title of the details tooltip.
 *
 * @param {boolean} open
 *     The prop that controls whether the tooltip is visible or not
 *
 * @param {CancelTooltip~onCancel} onCancel
 *     The callback triggered when the cancel button is pressed.
 * @param {CancelTooltip~onConfirm} onConfirm
 *     The callback triggered when the confirm button is pressed.
 *
 * @example
 *     <CancelTooltip
 *          title={this.im("pingid.policies.deleterule.label")}
 *          label={this.im("pingid.policies.deleterule.title")}
 *          open={this.state.isInviteOpen} onToggle={this._handleToggle}
 *          disabled={false}>
 *              <p>what ever callout content is</p>
 *     </CancelTooltip>
 **/

var CancelTooltip = React.createClass({
    propTypes: {
        "data-id": React.PropTypes.string,
        cancelButtonText: React.PropTypes.string.isRequired,
        confirmButtonText: React.PropTypes.string.isRequired,
        label: React.PropTypes.oneOfType([
            React.PropTypes.string,
            React.PropTypes.object
        ]),
        onConfirm: React.PropTypes.func.isRequired,
        onCancel: React.PropTypes.func.isRequired,
        positionClassName: React.PropTypes.string
    },

    render: function () {
        return (
            <DetailsTooltip
                data-id={this.props["data-id"] + "-cancel-tooltip"}
                className="cancel-tooltip"
                label={this.props.label}
                open={this.props.open}
                positionClassName={this.props.positionClassName}
                stateless={true}
                showClose={false}
                title={this.props.title}>
                <p data-id={this.props["data-id"] + "-cancel-tooltip-text"}>
                    {this.props.messageText}
                </p>
                <div className="buttons">
                    <button
                        className="cancel"
                        data-id={this.props["data-id"] + "-cancel-confirm-btn"}
                        onClick={this.props.onConfirm}>
                        {this.props.confirmButtonText}
                    </button>
                    <div>
                        <a
                            className="cancel"
                            data-id={this.props["data-id"] + "-cancel-deny-btn"}
                            onClick={this.props.onCancel}>
                            {this.props.cancelButtonText}
                        </a>
                    </div>
                </div>
            </DetailsTooltip>
        );
    }
});

module.exports = CancelTooltip;
