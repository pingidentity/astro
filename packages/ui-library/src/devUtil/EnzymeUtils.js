import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
