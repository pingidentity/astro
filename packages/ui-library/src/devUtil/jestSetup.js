
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
