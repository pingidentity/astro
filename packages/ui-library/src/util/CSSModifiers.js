export const lightInputs = "modifier_light-inputs";
export const darkInputs = "modifier_dark-inputs";
export const stripInputMargins = "modifier_strip-input-margins";

export default { lightInputs, darkInputs, stripInputMargins };
