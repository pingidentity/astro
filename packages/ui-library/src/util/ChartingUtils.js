import Color from "color";

const splitData = ({ id, data }) => data.map(point => ({ [id]: point }));

/**
 * Takes data from the format of [{ id: [data] }, { id2: [data2] }] and converts
 * it to the format of [[{ id: dataPoint }, { id2: dataPoint to }]]
 * All incoming arrays of data must be of the same length
 */
export const toRechartsDataFormat = dataSet => dataSet.reduce((acc, data) => {
    const split = splitData(data);
    return acc.length > 0
        ? acc.map((frame, index) => ({ ...frame, ...split[index] }))
        : split
    ;
}, []);

const addToExisting = (index, data, acc) => {
    const { id, data: currentData } = acc[index];
    return {
        data: [
            ...currentData,
            data,
        ],
        id
    };
};

/**
 * Takes data from the format of [[{ id: dataPoint }, { id2: dataPoint to }]] and converts
 * it to the format of [{ id: [data] }, { id2: [data2] }]
 */
export const fromRechartsDataFormat = dataSet => dataSet.reduce((acc, frame) =>
    Object.entries(frame).reduce((frameAcc, [id, data]) => {
        const idIndex = frameAcc.findIndex(entry => entry.id === id);
        return idIndex > -1
            ? [
                ...idIndex > 0 ? frameAcc.slice(0, idIndex) : [],
                addToExisting(idIndex, data, frameAcc),
                ...idIndex < frameAcc.length ? frameAcc.slice(idIndex + 1, frameAcc.length) : []
            ]
            : [...frameAcc, { id, data: [data] }];
    }, acc)
, []);

/**
 * Generates color props for chartes based on a given
 * "seed" color provided by the user
 */
export const generateTheme = (seedColor, data) => {
    const step = 360 / data.length;

    const dataColors = data.map((item, index) => {
        return {
            ...item,
            color: Color(seedColor).rotate(step * index).hex(),
        };
    });

    const contrastedColor = Color(seedColor).rotate(180).hex();

    return {
        highlightColor: contrastedColor,
        referenceLineColor: contrastedColor,
        referenceLabelColor: "#676D74",
        dataColors: dataColors,
    };
};

/**
 * Returns coordinates for drawing evenly spaced lines with a min and max offset
 */

export const getEvenLineCoords = (height, numLines = 5) => {
    //offsets are set to what recharts uses by default
    const minOffset = 20;
    const maxOffset = 5;

    const actualMin = 0 + minOffset;
    const actualMax = height - maxOffset;

    const range = actualMax - actualMin;

    const lines = Array.from( { length: numLines }, (v, i) => {
        return ( range / (numLines - 1) * (i) + minOffset);
    });

    return lines;
};
