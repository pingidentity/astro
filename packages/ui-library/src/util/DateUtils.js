/*eslint-disable valid-jsdoc*/

"use strict";
const moment = require("moment");
const dateFormat = "YYYY-MM-DD";
const timeFormat = "hh:MM:SSa";

const DateUtils = {
    /**
     * @alias module:util/Utils.formatDate
     * @desc Format the given timestamp in milliseconds as a YYYY-MM-DD date.
     *
     * @param {number} timestamp
     *     The timestamp (UNIX time * 1000)
     * @returns {string}
     *     Formatted date
     */
    formatDate: function (timestamp) {
        return moment.utc(timestamp).format(dateFormat);
    },

    /**
     * @alias module:util/Utils.formatTime
     * @desc Format the given timestamp in milliseconds as a YYYY-MM-DD hh:MM:SSa date.
     *
     * @param {number} timestamp
     *     The timestamp (UNIX time * 1000)
     * @returns {string}
     *     Formatted date+time
     */
    formatDateTime: function (timestamp) {
        return moment.utc(timestamp).format(`${dateFormat} ${timeFormat}`);
    },

    /**
     * @alias module:util/Utils.formatTime
     * @desc Format the given timestamp in milliseconds as a hh:MM:SSa time.
     *
     * @param {number} timestamp
     *     The timestamp (UNIX time * 1000)
     * @returns {string}
     *     Formatted time
     */
    formatTime: function (timestamp) {
        return moment.utc(timestamp).format(timeFormat);
    },
};

module.exports = DateUtils;