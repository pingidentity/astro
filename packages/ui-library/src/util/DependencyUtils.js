import React from "react";

/**
 * @module util/DependencyUtils
 * @desc Functions for showing nice errors when loading a component that has a requirement above our standard dependencies.
 */

/**
 * @alias module:util/DependencyUtils.usesStableContext
 *
 * @desc Throws an error if the React 16.3+ context API isn't available.
 *
 * @example
 *   usesStableContext()
 */

export const usesStableContext = (component) => {
    if (!React.createContext) {
        console.error(`${component} requires React 16.3`);
    }
};

/**
 * @alias module:util/DependencyUtils.usesFragments
 *
 * @desc Throws an error if React fragments aren't available.
 *
 * @example
 *   usesStableContext()
 */

export const usesFragments = (component) => {
    if (!React.Fragment) {
        console.error(`${component} requires React 16`);
    }
};
