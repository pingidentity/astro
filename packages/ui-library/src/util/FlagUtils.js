import PropTypes from "prop-types";

export const allFlags = [
    "pass-value",
    // v4 flags
    "p-stateful",
    "use-portal",
    "add-button-margin",
    "expandable-row-class",
    "fix-discard-button",
    "fixed-messages-constants",
    "true-default",
    "v4",
    "new-alert-modal",
];

export const flagsPropType = PropTypes.arrayOf(PropTypes.oneOf(allFlags));

export const getFlags = ({ props, context }) => props.flags || (context && context.flags) || [];

export const hasFlag = (component, flag) => {
    const flags = getFlags(component);
    return flags.includes(flag);
};
