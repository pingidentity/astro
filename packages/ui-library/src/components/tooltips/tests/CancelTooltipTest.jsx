import React from "react";
import { shallow } from "enzyme";
import { mountSnapshotDataIds } from "../../../devUtil/EnzymeUtils";
import { allFlags } from "../../../util/FlagUtils";
import CancelTooltip from "../CancelTooltip";

describe("CancelTooltip", () => {
    it("renders the component", () => {
        const component = shallow(
            <CancelTooltip
                cancelButtonText="OH NO GO BACK"
                confirmButtonText="What could go wrong?"
                flags={allFlags}
                onConfirm={jest.fn()}
                onCancel={jest.fn()}
            />
        );

        expect(component.exists()).toEqual(true);
    });

    it("data-id's don't change", () => {
        mountSnapshotDataIds(
            <CancelTooltip
                cancelButtonText="OH NO GO BACK"
                confirmButtonText="What could go wrong?"
                flags={allFlags}
                onConfirm={jest.fn()}
                onCancel={jest.fn()}
            />
        );
    });
});