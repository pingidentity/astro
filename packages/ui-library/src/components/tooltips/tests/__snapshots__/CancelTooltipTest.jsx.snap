// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CancelTooltip data-id's don't change 1`] = `
Array [
  "undefined-cancel-tooltip",
]
`;
